test/elixir_analyzer/exercise_test/common_checks/function_annotation_order_test.exs

```diff
@@ -0,0 +1,573 @@
+                type: :informative,
+                name: Constants.solution_function_annotation_order(),
+                comment: Constants.solution_function_annotation_order(),
+                params: %{
+                  expected: """
+                  @doc
+                  @spec b
+                  def b
+                  """,
+                  actual: """
+                  @spec b
+                  @doc
+                  def b
+                  """
+                }
+              }}
+           ]
+  end
+
+  test "returns multiple errors if it checks multiple times" do
+    ast =
+      quote do
+        defmodule Test do
+          @spec sum(number(), number()) :: number()
+          @doc "sum two numbers"
+          def sum(x, y), do: x + y
+
+          @spec subtract(number(), number()) :: number()
+          @doc "subtract two number"
+          def subtract(x, y), do: x - y
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == [
+             {:fail,
+              %Comment{
+                type: :informative,
+                name: Constants.solution_function_annotation_order(),
+                comment: Constants.solution_function_annotation_order(),
+                params: %{
+                  expected: """
+                  @doc
+                  @spec sum
+                  def sum
+                  """,
+                  actual: """
+                  @spec sum
+                  @doc
+                  def sum
+                  """
+                }
+              }},
+             {:fail,
+              %Comment{
+                type: :informative,
+                name: Constants.solution_function_annotation_order(),
+                comment: Constants.solution_function_annotation_order(),
+                params: %{
+                  expected: """
+                  @doc
+                  @spec subtract
+                  def subtract
+                  """,
+                  actual: """
+                  @spec subtract
+                  @doc
+                  def subtract
+                  """
+                }
+              }}
+           ]
+  end
+
+  test "spec between function definitions crashes" do
+    ast =
+      quote do
+        defmodule Test do
+          def test(x), do: x
+          @spec test(any()) :: any()
+          def test(x, y), do: x || y
+        end
+      end
+
+    assert FunctionAnnotationOrder.run(ast) == [
+             {:fail,
+              %Comment{
+                type: :informative,
+                name: Constants.solution_function_annotation_order(),
+                comment: Constants.solution_function_annotation_order(),
+                params: %{
+                  expected: """
+                  @doc
+                  @spec test
+                  def test
+                  """,
+                  actual: """
+                  @spec test
+                  @doc
+                  def test
+                  """
```

Comment:
I have an issue here. In many of the failing tests, the actual code has little to do with what we say is `actual`. In this particular example, there is no `@doc` at all, in the one below there is no `@spec`. This is guaranteed to confuse the students. 
Having an accurate `actual` is probably hard to do since the full picture is often broken between chunks, so my proposal is to remove the field altogether. We don't tell what the student did wrong specifically, we simply show them what is expected. Since they have the function name they will be able to figure it out. It will be a bit more work, but I believe leaving things as they are will be counter productive.
What do you think?

Reply:
I have been thinking about it, and actually it is not that hard, the chunks are per function name now, for example:
```elixir
        defmodule Test do
          def a(x \\ [])
          @doc ""
          @spec a(list()) :: atom()
          def a([]), do: :empty
          def a(_), do: :full

          @spec b
          def b

          @spec c
          def c
        end
```

Will generate the following chunk to `a` functions and attrs: `[:def, :doc, :spec, :def, :def]` So we could use this to define the `actual` field in a more precise way, and also define a sort function to use the sorted order in the `expected` field, i have not defined an algorithm to sort it yet, but i think it is not that hard.

What do you think about this?

Reply:
Sorting should be easy, but code like this
```elixir 
def a
@spec a
@doc ""
def a
```
Will be split into 2 different chunks right? `[:def, :spec, :doc]` and `[:def]`, so in general it might be a bit tricky to get the real full picture if you need to look at the previous/next chunk. Feel free to give it a go, but I'm wondering if it's worth the trouble.

Reply:
> @angelikatyborska I know you didn't follow the PR closely, but this is about communication with the students so I would like to have your opinion.

IMO we don't need any code blocks in the comment shown to the student. This text should be more than enough to figure out what the check is about:

>  Developers can choose the order of the `@doc` and `@spec` modules attributes, but the Elixir community convention is to use `@doc` first and `@spec` next to the function.

Reply:
Done! Since there is no code snippets, i limited the error to show only once